File: java-implementation/src/main/java/com/inventory/transport/Land.java
package com.inventory.transport;

import com.inventory.transport.TransportTypes.LandType;
import static com.inventory.transport.TransportTypes.LandTransportProperties;
import static com.inventory.transport.TransportTypes.LandRoutes;

// Represent the properties of land transport
public class Land extends Transport {
    private LandType type; // The type of land transport
    private String roadStages; // The predefined route stages for a specific land transport type

    // Constructs a Land transport instance with properties based on the properties
    public Land(LandType type) {

        // Call the base class constructor with speed and cost
        super(LandTransportProperties.get(type).getFirst(), LandTransportProperties.get(type).getSecond());
        this.type = type;
        this.roadStages = LandRoutes.get(type);
    }

    // Calculate delivery time
    @Override
    public double calculateDeliveryTime(double distance) {
        return distance / speed;
    }

    // Calculate the delivery cost
    @Override
    public double calculateDeliveryCost(double distance) {
        return distance * costPerKm;
    }

    // Return a string representation of the transport type
    @Override
    public String getTransportType() {
        return toString(type);
    }

    // Return the predefined land routes for this vessel
    @Override
    public String getRouteStages() {
        return roadStages;
    }

    // Converts the enum to a legible string
    private String toString(LandType type) {
        switch (type) {
            case Truck: return "Truck";
            case Car: return "Car";
            case Pickup: return "Pickup";
            default: return "Unknown";
        }
    }
}

File: cpp-implementation/src/transport/Land.cpp
#include "Transport.ih"

// Constructor for the Land class
Land::Land(LandType type)
    : Transport(LandTransportProperties.at(type).first, LandTransportProperties.at(type).second),
      type(type), roadStages(LandRoutes.at(type)) {

}

// Calculates the time taken to cover a distance at the object's speed.
double Land::calculateDeliveryTime() const {
    return distance / speed;
}

// Calculates the total cost of delivering over a distance.
double Land::calculateDeliveryCost() const {
    return distance * costPerKm;
}

// Converts the OceanType enum value to a readable string.
string Land::getTransportType() const {
    return toString(type);
}

// Returns a string describing the route stages for the specific ocean transport.
string Land::getRouteStages() const {
    return roadStages;
}

// Helper function to convert OceanType enum to string
string toString(LandType type) {
    switch (type) {
        case LandType::Truck: return "Truck";
        case LandType::Car: return "Car";
        case LandType::Pickup: return "Pickup";
        default: return "Unknown";
    }
}
